python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -0,0 +1,221 @@
+"""Node embedding optimizers for distributed training"""
+import abc
+from abc import abstractmethod
+import torch as th
+import torch.distributed as dist
+
+from ...dist_tensor import DistTensor
+from ...sparse_emb import NodeEmbedding
+from .utils import alltoallv_cpu, alltoall_cpu
+
+class DistSparseGradOptimizer(abc.ABC):
+    r''' The abstract dist sparse optimizer.
+
+    Note: dgl dist sparse optimizer only work with dgl.distributed.NodeEmbedding
+
+    Parameters
+    ----------
+    params : list of NodeEmbedding
+        The list of NodeEmbedding.
+    lr : float
+        The learning rate.
+    '''
+    def __init__(self, params, lr):
+        self._params = params
+        self._lr = lr
+        self._rank = None
+        self._world_size = None
+        self._shared_cache = {}
+        self._clean_grad = False
+        self._opt_meta = {}
+
+        for emb in params:
+            assert isinstance(emb, NodeEmbedding), \
+                'DGL DistSparseOptimizer only supports dgl.distributed.NodeEmbedding'
+
+            if self._rank is None:
+                self._rank = emb.rank
+                self._world_size = emb.world_size
+            else:
+                assert self._rank == emb.rank, \
+                    'MultiGPU rank for each embedding should be same.'
+                assert self._world_size == emb.world_size, \
+                    'MultiGPU world_size for each embedding should be same.'
+
+    def step(self):
+        ''' The step function.
+
+        The step function is invoked at the end of every batch to push the gradients
+        of the embeddings involved in a mini-batch to DGL's servers and update the embeddings.
+        '''
+        with th.no_grad():
+            local_indics = {emb.name: [] for emb in self._params}
+            local_grads = {emb.name: [] for emb in self._params}
+            device = th.device('cpu')
+            for emb in self._params:
+                name = emb._tensor.name
+                kvstore = emb._tensor.kvstore
+                trace = emb._trace
+                trainers_per_server = self._world_size // kvstore.num_servers
+
+                idics = [t[0] for t in trace]
+                grads = [t[1].grad.data for t in trace]
+                idics = th.cat(idics, dim=0)
+                grads = th.cat(grads, dim=0)
+                device = grads.device
+
+                # will send grad to each corresponding trainer
+                if self._world_size > 1:
+                    # get idx split from kvstore
+                    idx_split = kvstore.get_partid(name, idics)
+                    idx_split_size = []
+                    idics_list = []
+                    grad_list = []
+                    # split idx and grad first
+                    for i in range(kvstore.num_servers):
```

Comment:
I wonder if this is a good way to split the gradients. When we get more machines, this for loop may become slow.
KVStore actually tried to optimize this kind of data distribution.

Reply:
As everything is in GPU, it should be fast.
But, we can design some new API to replace this for loop.

Reply:
are there any profiling results on this piece of code?